File: Backend/src/main/java/com/espumablanca/backend/application/usecase/ClientService.java
package com.espumablanca.backend.application.usecase;

import com.espumablanca.backend.domain.model.dto.request.client.ClientRequest;
import com.espumablanca.backend.domain.model.dto.response.client.ClientDto;
import com.espumablanca.backend.domain.model.dto.response.client.ClientFullNameDto;
import com.espumablanca.backend.domain.model.dto.response.PageDto;
import org.springframework.data.domain.Pageable;


public interface ClientService {

    ClientFullNameDto create(ClientRequest request);

    ClientDto getById(Long id);

    PageDto getAll(String value, Pageable pageable);

    PageDto getAllEnabled(String value, Pageable pageable);

    ClientDto update(Long id, ClientRequest request);

    void delete(Long id);

    void enable(Long id);

    void disable(Long id);
}

File: Frontend/src/app/core/services/api/client.service.ts
import { Injectable, inject } from '@angular/core';
import { environment } from '../../../../environments/environment.development';
import { HttpClient, HttpParams } from '@angular/common/http';
import { ClientDto } from '../../models/client/response/client-dto.model';
import { PageDto } from '../../models/pageable/page-dto.model';
import { ClientRequest } from '../../models/client/request/client-request.model';

@Injectable({
  providedIn: 'root',
})
export class ClientService {
  private readonly endpoint = environment.apiUrl + '/protected/clients';
  private readonly http = inject(HttpClient);

  getAll(value?: string, page?: number) {
    let params = new HttpParams();
    if (value) {
      params = params.set('value', value);
    }
    if (page) {
      params = params.set('page', page.toString());
    }
    return this.http.get<PageDto>(this.endpoint, {
      params,
    });
  }

  getById(id: number) {
    return this.http.get<ClientDto>(`${this.endpoint}/${id}`);
  }

  createNew(request: ClientRequest) {
    return this.http.post<ClientDto>(this.endpoint, request);
  }

  update(id: number, request: ClientRequest) {
    return this.http.put<ClientDto>(`${this.endpoint}/${id}`, request);
  }

  delete(id: number) {
    return this.http.delete(`${this.endpoint}/${id}`);
  }

  enable(id: number) {
    return this.http.put(`${this.endpoint}/${id}/enable`, null);
  }

  disable(id: number) {
    return this.http.put(`${this.endpoint}/${id}/disable`, null);
  }
}
